### ISSUE TITLE
Add pilus damage cooldown on ejecting something 

### ISSUE THREAD

**Issue by hhyyrylainen** [Revolutionary-Games/Thrive#5139]
to make sure ejected cell cannot kill you immediately

**Comment by MalyshevArtem** [comment]
Can someone direct to the class/classes in which the cell ejection process is coded?

**Comment by CI09** [comment]
We use Entity Component System, meaning there are entities with components, managed by systems. In this case, `EngulfingSystem` has method `EjectEngulfable()`. Note that components sometimes also have helper methods but it's mostly systems that manage stuff in bigger picture.

**Comment by CI09** [comment]
Are you willing to help out with this issue @MalyshevArtem?

**Comment by MalyshevArtem** [comment]
Actually yes. But I don't know how much time it will take for me to do this. I want to spend some time after work when I'm free trying to solve it.

**Comment by MalyshevArtem** [comment]
@CI09 Is there a way to start the game at a specific point or maybe a way to set to my microbe specific parameters before the game starts, for example, add pili for damaging and ability to eject a cell so that I can test what I've coded, in other words, declare initial parameters for the game?

**Comment by CI09** [comment]
Probably there is (I'm not expert at developing here) by changing how base species is made but that's kinda hacky, would make auto evo evolve things off of it and maybe some other side issues.
For starting the game, for example in microbe stage, you can open godot and press the 🎬 looking button in top right corner while having microbestage.tscn opened, but I personally just run the game normally since it doesn't take long to get to freebuild.

**Comment by MalyshevArtem** [comment]
@CI09 Thanks for your help

**Comment by MalyshevArtem** [comment]
@hhyyrylainen Is it only related to pilus damage or other ways of damaging as well like toxins when a cell is ejected?

**Comment by CI09** [comment]
Toxins are already blocked from being exhausted when being in state other than normal so you don't have to do anything about it I think. Also he's on vacation until 22th so he might not answer.

**Comment by hhyyrylainen** [comment]
@MalyshevArtem this is purely for the pilus, because it can deal massive damage immediately killing the cell that just ejected something. There's various bug reports like this: https://community.revolutionarygamesstudio.com/t/massive-damage-when-ejecting-cell-with-pili/7559 complaining about the problem. Ejected cells don't immediately eject toxins so there isn't an equivalent problem there so just applying pilus damage cooldown is enough.

The way I'd approach this is by starting in `EngulfingSystem.PerformEjectionForceAndAttachedRemove` I'd get the `DamageCooldown` for the entity that ejected something and then adding like a bit of cooldown like: `damageCooldown.StartCooldown(Constants.PILUS_MIN_COOLDOWN)`. Then test if that helps. If it does all that's left is updating the system attributes to mark the new used component and making a PR. If it doesn't then perhaps a longer cooldown is needed or maybe the approach needs tweaking.

**Comment by MalyshevArtem** [comment]
@hhyyrylainen I've actually done something like this. I added two rows but in CompleteEjection:
`ref var cooldown = ref entity.Get<DamageCooldown>();`
`cooldown.CooldownRemaining = Constants.PILUS_MAX_COOLDOWN;`

It helped. But I'm gonna test it more and find the more suitable place for it.
I also feel it should work even if my cell is ejecting more than one cell because the cooldown is sort of global.

**Comment by MalyshevArtem** [comment]
@hhyyrylainen @CI09  Is there a good source to read about the concept of using attributes with systems for thread safety? I don't simply want to add `[WritesToComponent(typeof(DamageCooldown))]`
`[ReadsComponent(typeof(DamageCooldown))]`
not understanding how it works.

I've read https://github.com/Revolutionary-Games/Thrive/blob/master/doc/architecture.md and now I have some sense but it's not clear technically for me.

**Comment by hhyyrylainen** [comment]
I don't really know what more I could say without getting into the super specifics. https://github.com/Revolutionary-Games/Thrive/blob/master/doc/architecture.md#simulation-multithreading that basically explains it: there's a tool in Thrive for generating sequences of system executions that takes into account the attributes to ensure the sequence of system runs only runs non-conflicting systems in parallel. More details are in the code comments on the attributes and the `GenerateThreadedSystems` class.

So basically the attributes are the configuration data for `GenerateThreadedSystems` which calculates potential execution orders with parallelism and picks the best one it finds.

**Comment by MalyshevArtem** [comment]
@hhyyrylainen I updated my local Thrive repo to be up to date with the upstream repo. Then I decided to try to run GenerateThreadedSystems.tscn in Godot. It wrote:
```
Updating simulation class partial in src/microbe_stage/MicrobeWorldSimulation.generated.cs
Successfully handled. src/microbe_stage/MicrobeWorldSimulation.generated.cs has been updated
```
I think it's not bad because the script generates the right order anyway but why did it change the order when I'd not done any changes?

Then I also added my code with `[WritesToComponent(typeof(DamageCooldown))]` and ran GenerateThreadedSystems.tscn again. This time It didn't change anything. I suppose this attribute doesn't affect the order.

**Comment by hhyyrylainen** [comment]
It's entirely possible the run order happened to be correct (or there was some other system constraint that already caused the same effect).  The reason why the script is not fully consistent is that it uses a partially randomized approach to look for suitable system run orders, and only keeps the best one it finds. On my system with my 16 core CPU the script result is pretty much deterministic as the script runtime is long enough that the script basically always stumbles upon the optimal order it can find.

Once your PR is otherwise good to go, I can also check with the script on the PR to double verify that there isn't some issue that has popped up.

**Comment by hhyyrylainen** [comment]
Whoops the PR description didn't mark this as closing, so I'll do that manually now.

https://github.com/Revolutionary-Games/Thrive/pull/5318
